File: 133-clone-graph/clone-graph.js
/**
 * // Definition for a _Node.
 * function _Node(val, neighbors) {
 *    this.val = val === undefined ? 0 : val;
 *    this.neighbors = neighbors === undefined ? [] : neighbors;
 * };
 */

/**
 * @param {_Node} node
 * @return {_Node}
 */
var cloneGraph = function(node) {

    let node_new=node;
    if(node_new===null)
    {
        return null;
    }
    const vis = new Map();

    const dfs =(current_node)=>{
        if(vis.has(current_node))
        {
            return vis.get(current_node);
        }

        const new_clone=new Node(current_node.val);
        vis.set(current_node, new_clone);

        for(let neighbors of current_node.neighbors)
        {
            new_clone.neighbors.push(dfs(neighbors));
        }

        return new_clone;

    };

    return dfs(node_new);


    
};

File: 133-clone-graph/clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    private  Node dfs(Node node , HashMap<Integer,Node> vis ){
        if(vis.containsKey(node.val)){
            return vis.get(node.val);
        }
        Node newNode = new Node(node.val);
        vis.put(node.val,newNode);
        for(int i=0;i<node.neighbors.size();i++){
            newNode.neighbors.add(dfs(node.neighbors.get(i),vis));  
        }
        return newNode;
    }
    public Node cloneGraph(Node node) {
        if(node==null){
            return null;
        }
        return dfs(node , new HashMap<Integer,Node>());
    }
}